_posts/2021-09-06-terraform-uncommon-options.markdown

```diff
@@ -0,0 +1,336 @@
+  </a>
+</p>
+<br/>
+
+> "Once a new technology rolls over you, if you're not part of the steamroller, you're part of the road." - Stewart Brand
+
+<br/>
+
+The most frequently used Terraform commands, **plan**, **apply**, and **destroy** are the first things I learned when I started using Terraform. Primarily because those commands are a crucial part of Terraform’s cycle. Over time, as Terraform and its community grew, the features of Terraform grew in parallel. Now there are a lot more commands, options, arguments, and functions. Some are not used as often, and some do not even show when invoking the help menu.
+
+In this blog post, I want to talk about the not-so-very common commands, arguments, and use cases.
+
+
+## [Terraform Console]
+
+This command is my current personal fav, the **console** command. As per documentation it is most commonly used for experimenting with expressions and functions. As an added extra, if the command is invoked in a folder with an active Terraform state it will read it from the configured backend. Inspecting and experimenting with the state with Terraform's console can be fun!
+
+Getting to it is as simple as typing in the command line:
+
+```sh
+$ terraform console
+```
+
+Now we can start "playing" with our state, provided there is something in it that we can fiddle with.
+
+For the sake of this blog post let's create a `main.tf` file with the following contents. The file is also available in my [examples repo].
+
+```hcl
+locals {
+  map = {
+    a = 1
+    b = 2
+    c = 3
+  }
+}
+
+resource "local_file" "private_key" {
+  content  = jsonencode(local.map)
+  filename = "./map.json"
+}
+```
+Needless to say, executing `terraform init` is necessary to pull the plugins and initialize the configured backend if there is any.
+
+Afterwads running `terraform apply` against this template, a JSON encoded file with the content of the local variable `map` will be created in the same folder. That is easy enough.
+
+But what if I needed a YAML encoded file? I am also not sure how it will look alike. I prefer JSON, but I also like to have some comments.
+
+I can check out on the fly by firing up Terraform console in the same folder and just YAML encoding the local `map` variable:
+
+```sh
+$ terraform console
+> yamlencode(local.map)
+<<EOT
+"a": 1
+"b": 2
+"c": 3
+
+EOT
+```
+Voila! We see how our file will look like if we've used the `yamlencode` function instead.
+
+Why stop there, maybe I just need the value of the `b` key?
+
+```sh
+> yamlencode(local.map.b)
+<<EOT
+2
+...
+
+EOT
+```
+
+Swapping the keys and the values is also something that I might want?
+
+```sh
+> yamlencode({for k,v in local.map : v => k })
+<<EOT
+"1": "a"
+"2": "b"
+"3": "c"
+
+EOT
+```
+Or just "trying" out the `try` function to see how it exactly works.
+
+```sh
+> try(yamlencode(local.map.d), 4)
+4
+```
+
+I digress. The flexibility that the `terraform console` provides when testing, troubleshooting, or simply playing around is immense.
```

Comment:
I think what you mean here is "Case in point." "I digress" means that you're admitting you're getting distracted from your main argument, unless that's what you wanted to say and I didn't understand you.